File: go/problem0005/solution.go
// 2520 is the smallest number that can be divided by each of the numbers from
// 1 to 10 without any remainder.
// What is the smallest positive number that is evenly divisible by all of
// the numbers from 1 to 20?
package main

import (
	"eulerproject/utils/primes"
	"fmt"
)

func main() {
	limit := 20
	resultMap := make(map[int]int)
	p := primes.NewPrimes()
	for n := 2; n <= limit; n++ {
		for k, v := range primes.GetPrimeFactorization(n, p) {
			resultMap[k] = max(resultMap[k], v)
		}
	}
	fmt.Printf("Final result:\n%v\n", primes.ReducePrimeFactorization(resultMap))
}

File: python/problem0005/solution.py
from collections import Counter


def factors(value):
    primes = Counter()
    for x in range(2, value + 1):
        while not (value % x):
            primes[x] += 1
            value //= x
        if value == 1:
            break
    return primes


def run(limit):
    primes = Counter()
    for x in range(2, limit + 1):
        subprimes = factors(x)
        for prim, exp in subprimes.items():
            primes[prim] = max(primes[prim], exp)

    result = 1
    for prim, exp in primes.items():
        result *= prim**exp
    print(result)


if __name__ == "__main__":
    run(20)
